apps/api-extractor-model/src/mixins/ApiNameMixin.ts

```diff
@@ -94,6 +98,17 @@ export function ApiNameMixin<TBaseClass extends IApiItemConstructor>(baseClass:
 
       jsonObject.name = this.name;
     }
+
+    /** @internal */
+    public _getCanonicalReferenceName(): string {
+      // TODO: This is a temporary workaround for a limitation of the experimental DeclarationReference API.
+      // We will remove this when the final implementation is in place.
```

Comment:
@rbuckton when I call `DeclarationReference.addNavigationStep()` it seems to expect a possibly quoted name.  It runs the `Parser` to interpret it.  Whereas API Extractor needs to pass an unquoted name.  What's the best way to do that?  Then we can eliminate this `JSON.stringify()` workaround.

Reply:
It shouldn't require a quoted name, but it allows a quoted name. It uses quotemarks to surround a string containing reserved characters like `!`, `.`, `~`, `#`, etc. Is the issue that the name may contain quotemarks as part of the name, and `DeclarationReference` is misinterpreting it?

Reply:
API Extractor doesn't have complete support for ECMAScript symbols yet.  So when it sees something like this:
```ts
class X {
  [Symbol.iterator]() { }
}
```
...it documents it as if it had been declared like this:
```ts
class X {
  "[Symbol.iterator]"() { }
}
```
This isn't ideal, but it's better than nothing.

But when we run code like this, `addNavigationStep()` fails:
```ts
let ref = DeclarationReference.empty()
.addNavigationStep(Navigation.Members, "[Symbol.iterator]");
console.log(ref.toString());

// Fails with:
// Uncaught SyntaxError: Invalid DeclarationReference '[Symbol.iterator]'
//    at Parser.fail (eval at <anonymous> (main.js:526), <anonymous>:610:15)
```
Whereas this produces the expected result `"[Symbol.iterator]"`:
```ts
let ref = DeclarationReference.empty()
.addNavigationStep(Navigation.Members, JSON.stringify("[Symbol.iterator]"));
console.log(ref.toString());
```

It seems incorrect for `DeclarationReference.addNavigationStep()` to be invoking the parser.  It should accept an unquoted member name, right?

Reply:
The name `[Symbol.iterator]` (unquoted) should already be allowed by `DeclarationReference`. There's even a test to verify this ("parse bracketed component text").

Reply:
> ```
> // Fails with:
> // Uncaught SyntaxError: Invalid DeclarationReference '[Symbol.iterator]'
> //    at Parser.fail (eval at <anonymous> (main.js:526), <anonymous>:610:15)
> ```
Sounds like a bug in the `DeclarationReference` API.

Reply:
It looks like its a bug in the grammar. I intended `[Symbol.iterator]` to work, but neglected to allow `.` in *BracketedAtom*.

Reply:
`[Symbol.iterator]` should be fixed by https://github.com/microsoft/tsdoc/pull/173

Reply:
> Also, there is a `DeclarationReference.makeSafeComponent` that will call `JSON.stringify` on an invalid component.

@rbuckton But isn't it inefficient to invoke the parser every time `.addNavigationStep()` gets called?  

I can't imagine API Extractor ever calling `.addNavigationStep()` without first doing `DeclarationReference.makeSafeComponent()`.  So it seems inefficient to have to do work to quote the string and then invoke the parser to unquote it.  

Should we change `.addNavigationStep()` to assume its parameter is an unquoted plain identifier?